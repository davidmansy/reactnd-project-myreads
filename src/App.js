import React from 'react';
import * as BooksAPI from './BooksAPI';
import './styles/App.css';
import SearchBooks from './components/SearchBooks';
import ListBooks from './components/ListBooks';
import { Route } from 'react-router-dom';

class BooksApp extends React.Component {
  state = {
    books: [],
    isLoading: false,
    bookCatMap: {} //Built to speed up the data reconciliation with the search results
  };

  initBookCatMap(books) {
    return books.reduce((memo, book) => {
      memo[book.id] = book.shelf;
      return memo;
    }, {});
  }

  getAllBooks = () => {
    this.setState({ isLoading: true }, () => {
      BooksAPI.getAll().then(books => {
        this.setState({
          books,
          isLoading: false,
          bookCatMap: this.initBookCatMap(books)
        });
      });
    });
  };

  componentDidMount() {
    this.getAllBooks();
  }

  handleShelfChange = (book, shelf) => {
    this.setState({ isLoading: true }, () => {
      BooksAPI.update(book, shelf).then(() => {
        this.setState(currentState => {
          currentState.books.find(b => b.id === book.id).shelf = shelf;
          currentState.bookCatMap[book.id] = shelf;
          currentState.isLoading = false;
          return currentState;
        });
      });
    });
  };

  onHandleCloseSearch = () => {
    this.getAllBooks();
  };

  render() {
    const { books, isLoading, bookCatMap } = this.state;

    return (
      <div className="app">
        <Route
          exact
          path="/"
          render={() => (
            <ListBooks
              books={books}
              isLoading={isLoading}
              handleShelfChange={this.handleShelfChange}
            />
          )}
        />
        <Route
          path="/search"
          render={() => (
            <SearchBooks
              bookCatMap={bookCatMap}
              onHandleClose={this.onHandleCloseSearch}
            />
          )}
        />
      </div>
    );
  }
}

export default BooksApp;
