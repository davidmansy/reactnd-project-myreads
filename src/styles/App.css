@import url('https://fonts.googleapis.com/css?family=Amatic+SC:700|Pacifico');

html,
body,
.root {
  height: 100%;
}
body {
  line-height: 1.5;
}
body,
.app {
  background: rgb(0, 121, 191);
  color: white;
  font-family: 'Amatic SC', cursive;
}

/* main page */

.list-books-title {
  background: rgb(2, 106, 167);
  font-family: 'Pacifico', cursive;
  padding: 10px 0;
  text-align: center;
}

.list-books-title h1 {
  color: rgb(112, 176, 219);
  font-weight: 400;
  margin: 0;
}

.list-books-content {
  flex: 1;
  padding: 0 0 80px;
}

.bookshelves {
  display: flex;
}

.bookshelf {
  background-color: rgb(226, 228, 230);
  border-radius: 5px;
  color: #333;
  margin: 15px;
  min-width: 200px;
  padding: 10px;
}

.bookshelf-title {
  font-family: 'Pacifico', cursive;
  margin: 0;
}

.bookshelf-books {
  text-align: center;
}

.bookshelf-empty {
  font-size: 1.2em;
  padding: 10px;
}

.open-search {
  bottom: 25px;
  position: fixed;
  right: 25px;
}

.open-search a {
  background: #00c2e0;
  background-image: url('../icons/add.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 28px;
  border-radius: 50%;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  display: block;
  font-size: 0;
  height: 50px;
  width: 50px;
}

/* search page */

.search-books-bar {
  background-color: rgb(2, 106, 167);
  position: fixed;
  width: 100%;
  top: 0;
  left: 0;
  z-index: 5;
  display: flex;
  box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 0 6px rgba(0, 0, 0, 0.23);
  align-items: center;
}

.search-books-input-wrapper {
  flex: 1;
  border-radius: 5px;
}

.search-books-bar input {
  background-color: rgb(112, 176, 219);
  border: none;
  border-radius: 5px;
  color: rgb(2, 106, 167);
  font-size: 1.25em;
  margin: 10px;
  outline: none;
  padding: 10px;
  width: 98%;
}

.search-books-bar input::placeholder {
  color: rgb(2, 106, 167);
}

.search-books-bar input:focus {
  background-color: white;
  color: black;
}

.search-books-bar input:focus::placeholder {
  color: black;
}

.close-search {
  display: block;
  top: 20px;
  left: 15px;
  width: 50px;
  height: 53px;
  background-image: url('../icons/arrow-back.svg');
  background-position: center;
  background-repeat: no-repeat;
  background-size: 28px;
  font-size: 0;
}

.search-books-results {
  color: #333;
  padding: 80px 10px 20px;
}

/* common */
.spinner-container {
  align-items: center;
  display: flex;
  height: 600px;
  justify-content: center;
  width: 100%;
}

.spinner {
  width: 50px;
}

.technical-error {
  font-size: 2em;
}

/* books grid */

.books-grid {
  list-style-type: none;
  padding: 0;
  margin: 0;
  display: flex;
  justify-content: space-evenly;
  flex-wrap: wrap;
}

.books-grid li {
  padding: 10px;
  text-align: left;
}

.book {
  background-color: white;
  border-radius: 5px;
  height: 270px;
  padding: 10px;
  position: relative;
  width: 150px;
}

.book-active {
  background-color: #00c2e0;
  cursor: pointer;
}

.book-details-container {
  display: flex;
  flex-direction: column;
  height: 250px;
  overflow: auto;
  width: 130px;
}

.book-details-container-hidden {
  display: none;
}

.book-title {
  font-size: 1.2em;
  line-height: 1.2;
  margin-top: 10px;
}

.book-authors {
  color: #999;
  font-size: 0.9em;
}

.book-top {
  align-self: center;
  height: 200px;
}

.book-shelf-changer {
  position: absolute;
  right: -15px;
  bottom: -10px;
  width: 40px;
  height: 40px;
  border-radius: 50%;
  background: rgb(2, 106, 167);
  background-image: url('../icons/arrow-drop-down.svg');
  background-repeat: no-repeat;
  background-position: center;
  background-size: 20px;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
}

.book-shelf-changer select {
  width: 100%;
  height: 100%;
  opacity: 0;
  cursor: pointer;
}

/* book cover */

.book-cover {
  background: #eee;
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  border-radius: 5px;
}

.book-cover-title {
  padding: 20px 10px 0;
  text-align: center;
  font-size: 0.8em;
}
