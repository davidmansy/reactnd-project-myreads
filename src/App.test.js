import React from 'react';
import ReactDOM from 'react-dom';
import TestRenderer from 'react-test-renderer';
import App from './App';
import { MemoryRouter } from 'react-router-dom';

describe('App', () => {
  it('renders', () => {
    const container = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>,
      container
    );
    ReactDOM.unmountComponentAtNode(container);
  });

  test('snapshots', () => {
    const testRenderer = TestRenderer.create(
      <MemoryRouter initialEntries={['/']}>
        <App />
      </MemoryRouter>
    );
    let tree = testRenderer.toJSON();
    expect(tree).toMatchSnapshot();
    testRenderer.unmount();
  });
});

// it('renders without crashing', () => {
//   const wrapper = mount(
//     <MemoryRouter initialEntries={['/']}>
//       <App />
//     </MemoryRouter>
//   );
//   expect(wrapper.contains(<div className="bookshelves">)).toEqual(true);
//   console.log(wrapper.debug());
//   wrapper.unmount();
// });
