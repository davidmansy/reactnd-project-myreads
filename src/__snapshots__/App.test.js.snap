// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots 1`] = `
<div
  className="app"
>
  <div
    className="list-books"
  >
    <div
      className="list-books-title"
    >
      <h1>
        MyReads
      </h1>
    </div>
    <div
      className="list-books-content"
    >
      <div
        className="bookshelves"
      >
        <div
          className="spinner-container"
        >
          <img
            alt="Loading spinner"
            className="spinner"
            src="spinner.gif"
          />
        </div>
      </div>
    </div>
    <div
      className="open-search"
    >
      <a
        href="/search"
        onClick={[Function]}
      >
        Add a book
      </a>
    </div>
  </div>
</div>
`;
